File: 2026/02-February/07-Friday/solution.java
/**
 * Problem:
 * Given a string consisting only of characters 'a' and 'b',
 * determine the minimum number of deletions needed so that
 * all 'a' characters appear before any 'b' characters.
 *
 * In other words, the final string should be in the form:
 * "aaa...bbb..."
 *
 * Approach:
 * - Traverse the string from left to right.
 * - Keep track of how many 'b' characters have been seen so far.
 * - When an 'a' is found after a 'b', there are two choices:
 *   1) Delete this 'a'
 *   2) Delete all previous 'b's
 * - Choose the option with the minimum deletions.
 *
 * Time Complexity: O(n)
 * Space Complexity: O(1)
 */
class Solution {

    public int minimumDeletions(String s) {
        int bCount = 0;      // number of 'b's seen so far
        int deletions = 0;   // minimum deletions needed

        for (char ch : s.toCharArray()) {
            if (ch == 'b') {
                bCount++;
            } else { // ch == 'a'
                deletions = Math.min(deletions + 1, bCount);
            }
        }
        return deletions;
    }
}

File: 2026/02-February/07-Friday/solution.py
"""
Problem:
Given a string containing only 'a' and 'b', find the minimum number
of deletions required so that all 'a' characters come before any 'b'.

Approach:
- Iterate through the string once.
- Count how many 'b' characters have been seen.
- When an 'a' appears after 'b':
  - Either delete this 'a'
  - Or delete all previous 'b' characters
- Pick the option with fewer deletions.

Time Complexity: O(n)
Space Complexity: O(1)
"""

class Solution:
    def minimumDeletions(self, s: str) -> int:
        b_count = 0
        deletions = 0

        for ch in s:
            if ch == 'b':
                b_count += 1
            else:  # ch == 'a'
                deletions = min(deletions + 1, b_count)

        return deletions
